src/models/StyledComponent.js

```diff
@@ -50,20 +50,27 @@ export default (ComponentStyle: Function) => {
             this.setState({ theme, generatedClassName })
           })
         } else {
+          const theme = this.props.theme || {}
           const generatedClassName = this.generateAndInjectStyles(
-            this.props.theme || {},
+            theme,
             this.props
           )
-          this.setState({ generatedClassName })
+          this.setState({ theme, generatedClassName })
         }
       }
 
-      componentWillReceiveProps(nextProps: any) {
-        const generatedClassName = this.generateAndInjectStyles(
-          this.state.theme || this.props.theme,
-          nextProps
-        )
-        this.setState({ generatedClassName })
+      componentWillReceiveProps({ theme }: { theme: ?{ [key: string]: mixed } }) {
```

Comment:
@mxstbr This annotation is the same as `Theme` in `ThemeProvider`. I can export `Theme` type from `ThemeProvider` component and use it here. Then it would be `{ theme: ?Theme }`.